File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/NewExpressionSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public abstract class NewExpressionSyntax : ExpressionSyntax
	{
		internal SyntaxNode _attributeLists;

		public SyntaxToken NewKeyword => GetNewKeywordCore();

		public SyntaxList<AttributeListSyntax> AttributeLists => GetAttributeListsCore();

		internal NewExpressionSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal virtual SyntaxToken GetNewKeywordCore()
		{
			return new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.NewExpressionSyntax)base.Green)._newKeyword, base.Position, 0);
		}

		public NewExpressionSyntax WithNewKeyword(SyntaxToken newKeyword)
		{
			return WithNewKeywordCore(newKeyword);
		}

		internal abstract NewExpressionSyntax WithNewKeywordCore(SyntaxToken newKeyword);

		internal virtual SyntaxList<AttributeListSyntax> GetAttributeListsCore()
		{
			SyntaxNode red = GetRed(ref _attributeLists, 1);
			return new SyntaxList<AttributeListSyntax>(red);
		}

		public NewExpressionSyntax WithAttributeLists(SyntaxList<AttributeListSyntax> attributeLists)
		{
			return WithAttributeListsCore(attributeLists);
		}

		internal abstract NewExpressionSyntax WithAttributeListsCore(SyntaxList<AttributeListSyntax> attributeLists);

		public NewExpressionSyntax AddAttributeLists(params AttributeListSyntax[] items)
		{
			return AddAttributeListsCore(items);
		}

		internal abstract NewExpressionSyntax AddAttributeListsCore(params AttributeListSyntax[] items);
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/NewExpressionSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public MustInherit Class NewExpressionSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax
		Friend _attributeLists As SyntaxNode

		Public ReadOnly Property AttributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)
			Get
				Return Me.GetAttributeListsCore()
			End Get
		End Property

		Public ReadOnly Property NewKeyword As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return Me.GetNewKeywordCore()
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Public Function AddAttributeLists(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax
			Return Me.AddAttributeListsCore(items)
		End Function

		Friend MustOverride Function AddAttributeListsCore(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax

		Friend Overridable Function GetAttributeListsCore() As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)
			Return New SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)(MyBase.GetRed(Me._attributeLists, 1))
		End Function

		Friend Overridable Function GetNewKeywordCore() As Microsoft.CodeAnalysis.SyntaxToken
			Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.NewExpressionSyntax)._newKeyword, MyBase.Position, 0)
		End Function

		Public Function WithAttributeLists(ByVal attributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax
			Return Me.WithAttributeListsCore(attributeLists)
		End Function

		Friend MustOverride Function WithAttributeListsCore(ByVal attributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax

		Public Function WithNewKeyword(ByVal newKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax
			Return Me.WithNewKeywordCore(newKeyword)
		End Function

		Friend MustOverride Function WithNewKeywordCore(ByVal newKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NewExpressionSyntax
	End Class
End Namespace